pkg/controller/elasticsearch/shutdown/node.go

```diff
@@ -0,0 +1,156 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package shutdown
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/go-logr/logr"
+
+	esclient "github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
+)
+
+// NodeShutdown implements the shutdown.Interface with the Elasticsearch node shutdown API. It is not safe to call methods
+// on this struct concurrently from multiple go-routines.
+type NodeShutdown struct {
+	c           esclient.Client
+	typ         esclient.ShutdownType
+	reason      string
+	podToNodeID map[string]string
+	shutdowns   map[string]esclient.NodeShutdown
+	once        sync.Once
+	log         logr.Logger
+}
+
+var _ Interface = &NodeShutdown{}
+
+// NewNodeShutdown creates a new NodeShutdown struct restricted to one type of shutdown (typ); podToNodeID is mapping from
+// K8s Pod name to Elasticsearch node ID; reason is an arbitrary bit of metadata that will be attached to each node shutdown
+// request in Elasticsearch and can help tracking and auditing shutdown requests.
+func NewNodeShutdown(c esclient.Client, podToNodeID map[string]string, typ esclient.ShutdownType, reason string, l logr.Logger) *NodeShutdown {
+	return &NodeShutdown{
+		c:           c,
+		typ:         typ,
+		podToNodeID: podToNodeID,
+		reason:      reason,
+		log:         l,
+	}
+}
+
+func (ns *NodeShutdown) initOnce(ctx context.Context) error {
+	var err error
+	ns.once.Do(func() {
+		var r esclient.ShutdownResponse
+		r, err = ns.c.GetShutdown(ctx, nil)
+		if err != nil {
+			err = fmt.Errorf("while getting node shutdowns: %w", err)
+			return
+		}
+		shutdowns := map[string]esclient.NodeShutdown{}
+		for _, n := range r.Nodes {
+			ns.log.V(1).Info("Existing shutdown", "node_id", n.NodeID, "type", n.Type, "status", n.Status)
+			shutdowns[n.NodeID] = n
+		}
+		ns.shutdowns = shutdowns
+	})
+	return err
+}
+
+func (ns *NodeShutdown) lookupNodeID(podName string) (string, error) {
+	nodeID, exists := ns.podToNodeID[podName]
+	if !exists {
+		return "", fmt.Errorf("node %s currently not member of the cluster", podName)
+	}
+	return nodeID, nil
+}
+
+// ReconcileShutdowns retrieves ongoing shutdowns and based on the given node names either cancels or creates new
+// shutdowns.
+func (ns *NodeShutdown) ReconcileShutdowns(ctx context.Context, leavingNodes []string) error {
+	if err := ns.initOnce(ctx); err != nil {
+		return err
+	}
+	// cancel all ongoing shutdowns
+	if len(leavingNodes) == 0 {
+		return ns.Clear(ctx, nil)
```

Comment:
So the concern here is that we accumulate to many completed shutdowns in cluster state before we finally clean them up?

Reply:
Yes, my first understanding reading the comment is that it would cancel the shutdown for the nodes which are not in the slice:
> [...] based on the given node names either cancels or creates new shutdowns

But maybe I'm overthinking...